nemo_curator/utils/distributed_utils.py

```diff
@@ -421,34 +421,50 @@ def single_partition_write_with_filename(df, output_file_dir, output_type="jsonl
     else:
         success_ser = pd.Series([empty_partition])
 
-    if empty_partition:
-        filename = df.filename.iloc[0]
-        num_filenames = len(df.filename.unique())
-        if num_filenames > 1:
+    if not empty_partition:
+        filenames = df.filename.unique()
+        filenames = list(filenames.values_host) if is_cudf_type(df) else list(filenames)
+        num_files = len(filenames)
+        if num_files > 1 and output_type != "jsonl":
             raise ValueError(
-                f"More than one filename found in partition: {num_filenames}"
+                f"""
+                Found more than one filename({num_files}) in source partition which is not
+                supported for filetypes other than jsonl.
+                Consider using `files_per_partition=1` during read to ensure a single filename
+                source per partition of data.
+                """
             )
-        filename = Path(filename).stem
-        output_file_path = os.path.join(output_file_dir, filename)
-        if output_type == "jsonl":
-            output_file_path = output_file_path + ".jsonl"
-            if isinstance(df, pd.DataFrame):
-                df.to_json(
-                    output_file_path, orient="records", lines=True, force_ascii=False
-                )
+        for filename in filenames:
+            out_df = df[df.filename == filename] if num_files > 1 else df
+            filename = Path(filename).stem
+            output_file_path = os.path.join(output_file_dir, filename)
+            if output_type == "jsonl":
+                output_file_path = output_file_path + ".jsonl"
+                if isinstance(df, pd.DataFrame):
+                    out_df.to_json(
+                        output_file_path,
+                        orient="records",
+                        lines=True,
+                        force_ascii=False,
+                        mode="a",
+                    )
+                else:
+                    # See open issue here: https://github.com/rapidsai/cudf/issues/15211
+                    # df.to_json(
+                    #     output_file_path, orient="records", lines=True, engine="cudf", force_ascii=False
+                    # )
+                    out_df.to_json(
+                        output_file_path,
+                        orient="records",
+                        lines=True,
+                        force_ascii=False,
+                        mode="a",
+                    )
+            elif output_type == "parquet":
+                output_file_path = output_file_path + ".parquet"
+                out_df.to_parquet(output_file_path)
```

Comment:
We can expand to do this `to_parquet` with `partition_on` and that should probably just work .